File: python/tests/test_resample.py
"""Tests for standalone resampling: resample() and AudioResampler."""
import avioflow
import numpy as np
import pytest

# Only rounding of the rate ratio may move the count; the resampler tail is
# drained, so a larger gap means samples are being dropped. Mirrors
# SAMPLE_COUNT_TOLERANCE in avioflow/bin/resample-test.cpp.
SAMPLE_COUNT_TOLERANCE = 2


def make_sine(num_channels, num_samples, sample_rate, freq=440.0):
    """Generate a sine wave so output can be checked for amplitude, not just length."""
    t = np.arange(num_samples, dtype=np.float32) / sample_rate
    wave = np.sin(2 * np.pi * freq * t).astype(np.float32)
    return np.stack([wave] * num_channels)


def assert_sample_count(actual, expected, context):
    diff = actual - expected
    assert abs(diff) <= SAMPLE_COUNT_TOLERANCE, \
        f"{context}: expected ~{expected} samples, got {actual} (diff {diff})"


def test_one_shot_downsample():
    samples = make_sine(2, 44100, 44100)
    out = avioflow.resample(samples, 44100, 16000)

    assert out.shape[0] == 2, out.shape
    assert_sample_count(out.shape[1], 16000, "44100 -> 16000")
    assert out.dtype == np.float32, out.dtype

    # 440 Hz sits well below the 8 kHz Nyquist limit, so amplitude survives.
    peak = float(np.abs(out[0]).max())
    assert 0.9 < peak < 1.1, f"peak amplitude was {peak}"
    assert np.isfinite(out).all(), "output contains NaN or Inf"
    print("test_one_shot_downsample passed")


def test_upsample():
    samples = make_sine(2, 16000, 16000)
    out = avioflow.resample(samples, 16000, 44100)

    assert out.shape[0] == 2
    assert_sample_count(out.shape[1], 44100, "16000 -> 44100")
    print("test_upsample passed")


def test_equal_rates_pass_through():
    samples = make_sine(1, 4096, 16000)
    out = avioflow.resample(samples, 16000, 16000)

    assert out.shape == (1, 4096), out.shape
    print("test_equal_rates_pass_through passed")


def test_downmix_to_mono():
    samples = make_sine(2, 16000, 16000)
    out = avioflow.resample(samples, 16000, 16000, output_num_channels=1)

    assert out.shape == (1, 16000), out.shape
    print("test_downmix_to_mono passed")


def test_flush_recovers_tail():
    samples = make_sine(1, 44100, 44100)

    resampler = avioflow.AudioResampler(44100, 16000)
    body = resampler.process(samples)
    tail = resampler.flush()

    without_flush = body.shape[1]
    with_flush = without_flush + tail.shape[1]

    # The contract: only after flush() does the count match the rate ratio. How
    # much the resampler withholds is an internal detail, so assert the total.
    assert with_flush >= without_flush
    assert_sample_count(with_flush, 16000, "process + flush")
    assert abs(without_flush - 16000) > SAMPLE_COUNT_TOLERANCE, \
        f"expected process() alone to fall short, got {without_flush}"
    print(f"test_flush_recovers_tail passed (body={without_flush}, tail={tail.shape[1]})")


def test_chunked_matches_one_shot():
    samples = make_sine(1, 48000, 48000, freq=220.0)
    expected = avioflow.resample(samples, 48000, 16000)

    resampler = avioflow.AudioResampler(48000, 16000)
    parts = [resampler.process(samples[:, i:i + 1000])
             for i in range(0, samples.shape[1], 1000)]
    parts.append(resampler.flush())
    chunked = np.concatenate([p for p in parts if p.size], axis=1)

    # Filter state carries across chunks, so results match sample for sample.
    assert chunked.shape == expected.shape, (chunked.shape, expected.shape)
    max_diff = float(np.abs(chunked - expected).max())
    assert max_diff < 1e-6, f"max sample difference was {max_diff:e}"
    print(f"test_chunked_matches_one_shot passed (max diff {max_diff:e})")


def test_output_rate_and_channels_reported():
    resampler = avioflow.AudioResampler(44100, 16000)

    assert resampler.output_sample_rate == 16000
    # Channel count is unknown until the first chunk reveals it.
    assert resampler.output_num_channels == 0

    resampler.process(make_sine(2, 1000, 44100))
    assert resampler.output_num_channels == 2
    print("test_output_rate_and_channels_reported passed")


def test_accepts_foreign_arrays():
    # The point of the standalone API: resample audio avioflow did not decode.
    float64 = np.sin(np.linspace(0, 1, 44100)).reshape(1, -1)
    out = avioflow.resample(float64, 44100, 16000)
    assert out.shape[0] == 1
    assert_sample_count(out.shape[1], 16000, "float64 input")

    # A non-contiguous view must also work, via the forcecast copy.
    strided = make_sine(2, 88200, 44100)[:, ::2]
    out = avioflow.resample(strided, 22050, 16000)
    assert out.shape[0] == 2
    print("test_accepts_foreign_arrays passed")


def test_empty_flush():
    resampler = avioflow.AudioResampler(44100, 16000)
    tail = resampler.flush()
    assert tail.size == 0, tail.shape
    print("test_empty_flush passed")


@pytest.mark.parametrize(
    "call",
    [
        pytest.param(
            lambda samples: avioflow.resample(samples, 0, 16000),
            id="zero-input-rate",
        ),
        pytest.param(
            lambda samples: avioflow.resample(samples, 16000, -1),
            id="negative-output-rate",
        ),
        pytest.param(
            lambda samples: avioflow.resample(samples[0], 16000, 8000),
            id="one-dimensional-array",
        ),
        pytest.param(
            lambda samples: avioflow.AudioResampler(16000, 0),
            id="constructor-zero-rate",
        ),
    ],
)
def test_invalid_arguments(call):
    samples = make_sine(1, 100, 16000)
    with pytest.raises(ValueError):
        call(samples)


def test_rejects_channel_count_change():
    resampler = avioflow.AudioResampler(16000, 8000)
    resampler.process(make_sine(2, 100, 16000))
    with pytest.raises(ValueError):
        resampler.process(make_sine(1, 100, 16000))

File: nodejs/tests/test-resample.js
/**
 * Standalone resampling tests for the avioflow Node.js bindings:
 * resample() and the AudioResampler class.
 */
import avioflow from '../src/index.js'

// Only rounding of the rate ratio may move the count; the resampler tail is
// drained, so a larger gap means samples are being dropped. Mirrors
// SAMPLE_COUNT_TOLERANCE in avioflow/bin/resample-test.cpp.
const SAMPLE_COUNT_TOLERANCE = 2

/** Generate a sine wave so output can be checked for amplitude, not just length. */
function makeSine(numChannels, numSamples, sampleRate, freq = 440) {
    const wave = new Float32Array(numSamples)
    for (let i = 0; i < numSamples; i++) {
        wave[i] = Math.sin((2 * Math.PI * freq * i) / sampleRate)
    }
    return Array.from({ length: numChannels }, () => Float32Array.from(wave))
}

function assertSampleCount(actual, expected, context) {
    const diff = actual - expected
    if (Math.abs(diff) > SAMPLE_COUNT_TOLERANCE) {
        throw new Error(`${context}: expected ~${expected} samples, got ${actual} (diff ${diff})`)
    }
}

function peak(channel) {
    let max = 0
    for (const value of channel) max = Math.max(max, Math.abs(value))
    return max
}

function assertThrows(description, fn) {
    try {
        fn()
    } catch {
        return
    }
    throw new Error(`${description} should have thrown`)
}

function testOneShotDownsample() {
    console.log('\n=== Test: One-shot downsample 44100 -> 16000 ===')
    const input = makeSine(2, 44100, 44100)
    const output = avioflow.resample(input, 44100, 16000)

    if (output.length !== 2) throw new Error(`expected 2 channels, got ${output.length}`)
    assertSampleCount(output[0].length, 16000, '44100 -> 16000')
    if (output[1].length !== output[0].length) throw new Error('channel lengths differ')

    // 440 Hz sits well below the 8 kHz Nyquist limit, so amplitude survives.
    const amplitude = peak(output[0])
    console.log(`  ${output.length} ch x ${output[0].length} samples, peak ${amplitude.toFixed(3)}`)
    if (amplitude < 0.9 || amplitude > 1.1) throw new Error(`peak was ${amplitude}`)
    if (!output[0].every(Number.isFinite)) throw new Error('output contains NaN or Inf')
}

function testUpsample() {
    console.log('\n=== Test: Upsample 16000 -> 44100 ===')
    const output = avioflow.resample(makeSine(2, 16000, 16000), 16000, 44100)
    if (output.length !== 2) throw new Error('channel count changed')
    assertSampleCount(output[0].length, 44100, '16000 -> 44100')
    console.log(`  ${output[0].length} samples`)
}

function testEqualRatesPassThrough() {
    console.log('\n=== Test: Equal rates pass through ===')
    const output = avioflow.resample(makeSine(1, 4096, 16000), 16000, 16000)
    if (output.length !== 1 || output[0].length !== 4096) {
        throw new Error(`expected 1x4096, got ${output.length}x${output[0].length}`)
    }
    console.log('  unchanged')
}

function testDownmixToMono() {
    console.log('\n=== Test: Stereo to mono ===')
    const output = avioflow.resample(makeSine(2, 16000, 16000), 16000, 16000, 1)
    if (output.length !== 1 || output[0].length !== 16000) {
        throw new Error(`expected 1x16000, got ${output.length}x${output[0].length}`)
    }
    console.log('  1 channel')
}

function testFlushRecoversTail() {
    console.log('\n=== Test: flush() recovers the tail ===')
    const resampler = new avioflow.AudioResampler({
        inputSampleRate: 44100,
        outputSampleRate: 16000,
    })

    const body = resampler.process(makeSine(1, 44100, 44100))
    const tail = resampler.flush()
    const withoutFlush = body[0].length
    const withFlush = withoutFlush + tail[0].length

    console.log(`  process()=${withoutFlush}, flush()=${tail[0].length}, total=${withFlush}`)

    // The contract: only after flush() does the count match the rate ratio.
    if (withFlush < withoutFlush) throw new Error('flush lost samples')
    assertSampleCount(withFlush, 16000, 'process + flush')
    if (Math.abs(withoutFlush - 16000) <= SAMPLE_COUNT_TOLERANCE) {
        throw new Error(`expected process() alone to fall short, got ${withoutFlush}`)
    }
}

function testChunkedMatchesOneShot() {
    console.log('\n=== Test: Chunked equals one-shot ===')
    const sampleRate = 48000
    const input = makeSine(1, sampleRate, sampleRate, 220)
    const expected = avioflow.resample(input, sampleRate, 16000)[0]

    const resampler = new avioflow.AudioResampler({
        inputSampleRate: sampleRate,
        outputSampleRate: 16000,
    })

    const parts = []
    for (let i = 0; i < sampleRate; i += 1000) {
        parts.push(resampler.process([input[0].subarray(i, i + 1000)])[0])
    }
    parts.push(resampler.flush()[0])

    let total = 0
    for (const part of parts) total += part.length
    const chunked = new Float32Array(total)
    let offset = 0
    for (const part of parts) {
        chunked.set(part, offset)
        offset += part.length
    }

    // Filter state carries across chunks, so results match sample for sample.
    if (chunked.length !== expected.length) {
        throw new Error(`length differs: ${chunked.length} vs ${expected.length}`)
    }
    let maxDiff = 0
    for (let i = 0; i < chunked.length; i++) {
        maxDiff = Math.max(maxDiff, Math.abs(chunked[i] - expected[i]))
    }
    console.log(`  ${chunked.length} samples, max diff ${maxDiff.toExponential(2)}`)
    if (maxDiff > 1e-6) throw new Error(`max difference was ${maxDiff}`)
}

function testOutputRateAndChannelsReported() {
    console.log('\n=== Test: Output rate and channel count ===')
    const resampler = new avioflow.AudioResampler({
        inputSampleRate: 44100,
        outputSampleRate: 16000,
    })

    if (resampler.outputSampleRate() !== 16000) throw new Error('wrong output rate')
    // Channel count is unknown until the first chunk reveals it.
    if (resampler.outputNumChannels() !== 0) throw new Error('expected 0 channels before process')

    resampler.process(makeSine(2, 1000, 44100))
    if (resampler.outputNumChannels() !== 2) throw new Error('expected 2 channels after process')
    console.log('  16000 Hz, 0 channels before process, 2 after')
}

function testEmptyFlush() {
    console.log('\n=== Test: flush() before any process() ===')
    const resampler = new avioflow.AudioResampler({
        inputSampleRate: 44100,
        outputSampleRate: 16000,
    })
    const tail = resampler.flush()
    if (tail.length !== 0) throw new Error(`expected no channels, got ${tail.length}`)
    console.log('  handled')
}

function testInvalidArguments() {
    console.log('\n=== Test: Invalid arguments ===')
    const input = makeSine(1, 100, 16000)

    assertThrows('zero input rate', () => avioflow.resample(input, 0, 16000))
    assertThrows('negative output rate', () => avioflow.resample(input, 16000, -1))
    assertThrows('non-array samples', () => avioflow.resample('nope', 16000, 8000))
    assertThrows('non-Float32Array channel', () => avioflow.resample([[1, 2, 3]], 16000, 8000))
    assertThrows('constructor zero rate', () => new avioflow.AudioResampler({
        inputSampleRate: 16000,
        outputSampleRate: 0,
    }))

    const resampler = new avioflow.AudioResampler({
        inputSampleRate: 16000,
        outputSampleRate: 8000,
    })
    resampler.process(makeSine(2, 100, 16000))
    assertThrows('channel count change', () => resampler.process(makeSine(1, 100, 16000)))
    console.log('  all rejected')
}

function main() {
    try {
        testOneShotDownsample()
        testUpsample()
        testEqualRatesPassThrough()
        testDownmixToMono()
        testFlushRecoversTail()
        testChunkedMatchesOneShot()
        testOutputRateAndChannelsReported()
        testEmptyFlush()
        testInvalidArguments()
        console.log('\nAll resample tests passed!')
    } catch (error) {
        console.error(`\nTest failed: ${error.message}`)
        process.exit(1)
    }
}

main()
